### ISSUE TITLE
USB Driver / Interface

### ISSUE THREAD

**Issue by broberts-nybble** [andrey-ushakov/esc_pos_printer#10]
Hi, 

Thank you for this amazing library firstly, secondly I am going to pull it tomorrow as I would like to add the usb printing facility.

Given that it is your work, do you have any guidance?

I was looking at using https://pub.dev/packages/usb_serial, then sending the same commands over the usb serial port? are there any specifics that you are aware of that we should take into consideration.

Once we have it working, we will give it right back to you :)

**Comment by andrey-ushakov** [comment]
Hi,
Thank you for your feedback, I really appreciate it!
I didn't have time yet to dig into USB support but it looks like it's feasible using usb_serial library.
Some people already asked about the USB support, so I think it could be really useful to add this feature!

**Comment by broberts-nybble** [comment]
Hi Andrey,

Just to let you know, we have completed the usb driver for this, we are busy testing. 

It was really simple enough to get going with very few changes.

**Comment by andrey-ushakov** [comment]
Hi Brandon,

Sounds great! Thank you for your work!
I'm waiting for your pull request to test it.

**Comment by andrey-ushakov** [comment]
Hi Brandon,

Are you ready to make a pull request? I could include your changes into the next release.

**Comment by kevj121** [comment]
I would be really interested in testing this too? We were looking at embarking on a similar integration.

**Comment by n0tc0ding** [comment]
@broberts-nybble can you share the usb part for this please?

**Comment by AyushBherwani1998** [comment]
Any updates on the support for USB Driver? I would like to work if @andrey-ushakov you can provide me some guidance.

**Comment by andrey-ushakov** [comment]
Hi @AyushBherwani1998 ,

That will be really helpful!

I expect it shouldn't be too hard to develop using this [usb_serial library](https://pub.dev/packages/usb_serial) (android only) but I don't have an appropriate device to which I can connect my POS printer and test.

So it would be great if you could play a bit with an Android device, POS printer and this usb lib. Basically all we need to do is to send same POS commands over USB.

For example, you can create a simple demo app where user will choose an USB device (his POS printer) from a list and after that you can print a test ticket sending commands via USB.

The most interesting part would be how to create a class architecture to support USB and WiFi (and probably Bluetooth in future) connections.

Probably we can create a low-level (abstract) class `PrinterAbstract` that will generate POS commands and 2 classes - `PrinterWifi` and `PrinterUsb` that will send those commands over Wifi or Usb.

**Comment by AyushBherwani1998** [comment]
> 
> 
> Hi @AyushBherwani1998 ,
> 
> That will be really helpful!
> 
> I expect it shouldn't be too hard to develop using this [usb_serial library](https://pub.dev/packages/usb_serial) (android only) but I don't have an appropriate device to which I can connect my POS printer and test.
> 
> So it would be great if you could play a bit with an Android device, POS printer and this usb lib. Basically all we need to do is to send same POS commands over USB.
> 
> For example, you can create a simple demo app where user will choose an USB device (his POS printer) from a list and after that you can print a test ticket sending commands via USB.
> 
> The most interesting part would be how to create a class architecture to support USB and WiFi (and probably Bluetooth in future) connections.
> 
> Probably we can create a low-level (abstract) class `PrinterAbstract` that will generate POS commands and 2 classes - `PrinterWifi` and `PrinterUsb` that will send those commands over Wifi or Usb.

Hey, I'm trying to connect the thermal printer but it always throws a Platform Exception. I saw the issue was closed by @broberts-nybble 

@broberts-nybble  A little help would be appreciated.

**Comment by broberts-nybble** [comment]
> > Hi @AyushBherwani1998 ,
> > That will be really helpful!
> > I expect it shouldn't be too hard to develop using this [usb_serial library](https://pub.dev/packages/usb_serial) (android only) but I don't have an appropriate device to which I can connect my POS printer and test.
> > So it would be great if you could play a bit with an Android device, POS printer and this usb lib. Basically all we need to do is to send same POS commands over USB.
> > For example, you can create a simple demo app where user will choose an USB device (his POS printer) from a list and after that you can print a test ticket sending commands via USB.
> > The most interesting part would be how to create a class architecture to support USB and WiFi (and probably Bluetooth in future) connections.
> > Probably we can create a low-level (abstract) class `PrinterAbstract` that will generate POS commands and 2 classes - `PrinterWifi` and `PrinterUsb` that will send those commands over Wifi or Usb.
> 
> Hey, I'm trying to connect the thermal printer but it always throws a Platform Exception. I saw the issue was closed by @broberts-nybble
> 
> @broberts-nybble A little help would be appreciated.

Hey @AyushBherwani1998, basically the issue was that the device was not set to virtual com, what needed to happen was I had to plug it into my PC, open the sample app they generally provide and set it's comm protocol to Virtual COM, then it connected and did not crash.

**Comment by nguyenphuochai1998** [comment]
> > > Hi @AyushBherwani1998 ,
> > > That will be really helpful!
> > > I expect it shouldn't be too hard to develop using this [usb_serial library](https://pub.dev/packages/usb_serial) (android only) but I don't have an appropriate device to which I can connect my POS printer and test.
> > > So it would be great if you could play a bit with an Android device, POS printer and this usb lib. Basically all we need to do is to send same POS commands over USB.
> > > For example, you can create a simple demo app where user will choose an USB device (his POS printer) from a list and after that you can print a test ticket sending commands via USB.
> > > The most interesting part would be how to create a class architecture to support USB and WiFi (and probably Bluetooth in future) connections.
> > > Probably we can create a low-level (abstract) class `PrinterAbstract` that will generate POS commands and 2 classes - `PrinterWifi` and `PrinterUsb` that will send those commands over Wifi or Usb.
> > 
> > 
> > Hey, I'm trying to connect the thermal printer but it always throws a Platform Exception. I saw the issue was closed by @broberts-nybble
> > @broberts-nybble A little help would be appreciated.
> 
> Hey @AyushBherwani1998, basically the issue was that the device was not set to virtual com, what needed to happen was I had to plug it into my PC, open the sample app they generally provide and set it's comm protocol to Virtual COM, then it connected and did not crash.
Can you give me the code ?? plz

**Comment by nguyenphuochai1998** [comment]
> > > Hi @AyushBherwani1998 ,
> > > That will be really helpful!
> > > I expect it shouldn't be too hard to develop using this [usb_serial library](https://pub.dev/packages/usb_serial) (android only) but I don't have an appropriate device to which I can connect my POS printer and test.
> > > So it would be great if you could play a bit with an Android device, POS printer and this usb lib. Basically all we need to do is to send same POS commands over USB.
> > > For example, you can create a simple demo app where user will choose an USB device (his POS printer) from a list and after that you can print a test ticket sending commands via USB.
> > > The most interesting part would be how to create a class architecture to support USB and WiFi (and probably Bluetooth in future) connections.
> > > Probably we can create a low-level (abstract) class `PrinterAbstract` that will generate POS commands and 2 classes - `PrinterWifi` and `PrinterUsb` that will send those commands over Wifi or Usb.
> > 
> > 
> > Hey, I'm trying to connect the thermal printer but it always throws a Platform Exception. I saw the issue was closed by @broberts-nybble
> > @broberts-nybble A little help would be appreciated.
> 
> Hey @AyushBherwani1998, basically the issue was that the device was not set to virtual com, what needed to happen was I had to plug it into my PC, open the sample app they generally provide and set it's comm protocol to Virtual COM, then it connected and did not crash.

Can you show me code plz!!

**Comment by ccfiel** [comment]
@broberts-nybble any updates in your PR?

**Comment by AyushBherwani1998** [comment]
I was not able to achieve the functionality using [usb_serial library](https://pub.dev/packages/usb_serial). Instead, I used another package [esc-pos-printer-flutter](https://github.com/marcusfelix/esc-pos-printer-flutter) with combination with this package and made an interface to achieve usb and bluetooth. If you are using [esc-pos-printer-flutter](https://github.com/marcusfelix/esc-pos-printer-flutter) package you need to add other functionalities by yourself.

**Comment by ccfiel** [comment]
esc-pos-printer-flutter this is for usb. What package did you use for bluetooth? @AyushBherwani1998

**Comment by AyushBherwani1998** [comment]
> 
> 
> esc-pos-printer-flutter this is for usb. What package did you use for bluetooth? @AyushBherwani1998

For bluetooth I used [blue_thermal_printer](https://pub.dev/packages/blue_thermal_printer). It has all the print options you require.

**Comment by ccfiel** [comment]
@AyushBherwani1998 I think this package does not support ios? Have you tried it in ios?

**Comment by AyushBherwani1998** [comment]
> 
> 
> @AyushBherwani1998 I think this package does not support ios? Have you tried it in ios?

Yes, this package only supports Android as of now. You will need to modify the package to support the iOS using `PlatformChannel`.

**Comment by ccfiel** [comment]
Thanks @AyushBherwani1998

**Comment by komputerboy** [comment]
any update on this?
